tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc

```diff
@@ -156,12 +93,62 @@ TfLiteStatus AverageEvalInt8(TfLiteContext* context, const TfLiteNode* node,
   op_params.padding_values.width = data->padding.width;
   op_params.quantized_activation_min = activation_min;
   op_params.quantized_activation_max = activation_max;
-  reference_integer_ops::AveragePool(
-      op_params, GetTensorShape(input), GetTensorData<int8_t>(input),
-      GetTensorShape(output), GetTensorData<int8_t>(output));
 
-#endif
-  return kTfLiteOk;
+  if (input->type == kTfLiteUInt8) {
+    reference_ops::AveragePool(
+        op_params, GetTensorShape(input), GetTensorData<uint8_t>(input),
+        GetTensorShape(output), GetTensorData<uint8_t>(output));
+  } else {
+    TFLITE_DCHECK_LE(activation_min, activation_max);
+
+    RuntimeShape input_shape = GetTensorShape(input);
+    TFLITE_DCHECK_EQ(input_shape.DimensionsCount(), 4);
+
+    RuntimeShape output_shape = GetTensorShape(output);
+    TFLITE_DCHECK_EQ(output_shape.DimensionsCount(), 4);
+
+    const int depth = MatchingDim(input_shape, 3, output_shape, 3);
+
+    cmsis_nn_dims input_dims;
+    input_dims.n = 1;
+    input_dims.h = input_shape.Dims(1);
+    input_dims.w = input_shape.Dims(2);
+    input_dims.c = depth;
+
+    cmsis_nn_dims output_dims;
+    output_dims.n = 1;
+    output_dims.h = output_shape.Dims(1);
+    output_dims.w = output_shape.Dims(2);
+    output_dims.c = depth;
+
+    cmsis_nn_pool_params pool_params;
+    pool_params.stride.h = params->stride_height;
+    pool_params.stride.w = params->stride_width;
+    pool_params.padding.h = data->padding.height;
+    pool_params.padding.w = data->padding.width;
+    pool_params.activation.min = activation_min;
+    pool_params.activation.max = activation_max;
+
+    cmsis_nn_dims filter_dims;
+    filter_dims.n = 1;
+    filter_dims.h = params->filter_height;
+    filter_dims.w = params->filter_width;
+    filter_dims.c = 1;
+
+    auto* buffer_idx = reinterpret_cast<int*>(node->user_data);
```

Comment:
Prefer using OpData properly here, adding buffer_index to OpData and using data->buffer_index like in fully_connected.cc

Reply:
Sure. As part of the change, I moved the CalcOpData() and checks to the prepare phase similar to fully_connected and conv's as well.

Reply:
Sounds good.  I don't see the changes here yet - I assume this should become OpData* data = node->user_data, then using data->buffer_idx.